_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+passing the `const sus::Vec&` directly.
+
+But lastly, we have the call to `take_slice_ref()`, which now differs significantly from the
+equivalent `take_span_ref()` with the standard library types:
+```asm
+lea    rax,[rbp-0x70]
+mov    rdi,rax
+call   404bf0 <sus::Vec<int>::operator sus::Slice<int>&() &>
+mov    rdi,rax
+call   403fd1 <take_slice_ref(sus::Slice<int> const&)>
+```
+Now we're down to 5 instructions. We call the conversion operator to `sus::Slice&`, which returns
+_a pointer_ that we store, and then call `take_slice_ref()` with it.
+
+But why is this operator returning _a pointer_, instead of a `sus::Slice` by value? The trick here
+is that `sus::Vec` is implemented by holding a `sus::SliceMut` inside it, which holds a
+`sus::Slice` inside it, which is where the pointer/len pair is finally found. This means that `Vec`
+can convert to a `SliceMut` and `Slice` without any constructor being invoked.
+
+### With optimizations
+
+Ok let's turn some optimizations back on and see what we get.
+
+Calling `take_vector(const std::vector&)` grabs the pointer and invokes `call`.
+```asm
+lea    rdi,[rsp+0x10]
+call   4035e6 <take_vector(std::vector<int, std::allocator<int> > const&)>
+```
+
+Calling `take_span(std::span)` inlined the `std::span` constructor for us, which sets to values and
+then invokes `call`.
+```asm
+mov    rdi,rbx
+mov    esi,0x7
+call   403719 <take_span(std::span<int const, 18446744073709551615ul>)>
+```
+
+Calling `take_span_ref(const std::span&)` is worse, as it constructs a `std::span` inline, then has
+to take its address. No doubt that's where we get the guidance from to pass it by value.
+```asm
+mov    QWORD PTR [rsp+0x30],rbx
+mov    QWORD PTR [rsp+0x38],0x7
+lea    rdi,[rsp+0x30]
+call   4037ab <take_span_ref(std::span<int const, 18446744073709551615ul> const&)>
+```
+
+Calling `take_vec(const sus::Vec&)` looks similar to receiving `const std::vector&`, as it grabs
+the pointer and invokes `call`.
+```asm
+mov    rdi,rsp
+call   403840 <take_vec(sus::Vec<int> const&)>
+```
+
+Calling `take_slice(sus::Slice)` looks similar to receiving `std::span`, inlining the construction
+of a `sus::Slice`, then passing it by value.
+```asm
+mov    rdi,QWORD PTR [rsp]
+mov    rsi,QWORD PTR [rsp+0x8]
+call   4038e3 <take_slice(sus::Slice<int>)>
+```
+
+Calling `take_slice_ref(const sus::Slice&)` is where things differ in a meaningful way.
+```asm
+mov    rdi,rsp
+call   403981 <take_slice_ref(sus::Slice<int> const&)>
+```
+
+The call to `take_slice_ref()` uses the exact same instructions as the call to `take_vec()`. *This
+means we are able to use a bounded view type (`sus::Slice`) as a function parameter without taking
+a binary size and performance cost at every function call that moves from an owning type to a view
+type*. And that's something that we can't do with the standard library types.
+
+## Library choices that got us here
+
+There's a couple differences in Subspace that collude to get us to this point.

```

Comment:
Maybe "work together" rather than "collude"? "Collude" comes with a connotation of wrongdoing.